### PR TITLE
Noita: Extra Life change

### PR DESCRIPTION
**Pull request by ScipioWright**
## What is this fixing or adding?
Primarily, greatly reducing the rate of extra life perks. On the client side, making it so extra life perks redeliver on new game, making them a persistent buff rather than a buff for the current run. Should smooth out progression and make later runs more consistent.

And while in there, modified a few other rates, added a minimum amount of items to generate for a few things, and removed unnecessary reverse region connections.

## How was this tested?
Generated, ran the modified client.

## If this makes graphical changes, please attach screenshots.
N/A

### CODE REVIEW

**Comment by heinermann** [comment]
For region connections I thought it should represent how the game represents them. For example taking the Holy Mountain portal is a one way direction for us, other games have pits and stuff you fall down. Maybe a primary maintainer can clarify.

**Comment by ScipioWright** [comment]
> For region connections I thought it should represent how the game represents them. For example taking the Holy Mountain portal is a one way direction for us, other games have pits and stuff you fall down. Maybe a primary maintainer can clarify.

A discussion came up in the Tunic PR -- basically it's unnecessary reverse connections -- that is, Entrance objects that don't need to be made because they will never be the only way you can get to a specific spot. So like, any Entrance from a place to the Forest is completely unnecessary since you start there.

**Comment by heinermann** [comment]
Ok, we can probably change the way we generate regions later to get rid of the empty sets in the mapping.

**Review by heinermann — APPROVED** [review]

File: worlds/noita/Regions.py [hunk 1]
```diff
@@ -82,66 +82,64 @@ def create_all_regions_and_connections(multiworld: MultiWorld, player: int) -> N
...
     "Underground Jungle Holy Mountain": {"Underground Jungle"},
-    "Underground Jungle": {"Underground Jungle Holy Mountain", "Dragoncave", "Overgrown Cavern", "Vault Holy Mountain",
-                           "Lukki Lair"},
-    "Dragoncave": {"Underground Jungle"},
-    "Lukki Lair": {"Underground Jungle", "Snow Chasm", "Frozen Vault"},
-    "Snow Chasm": {},
+    "Underground Jungle": {"Dragoncave", "Overgrown Cavern", "Vault Holy Mountain", "Lukki Lair", "Snow Chasm"},
+    "Dragoncave": set(),
+    "Lukki Lair": set(),
+    "Snow Chasm": set(),
```
Comment (heinermann) [inline worlds/noita/Regions.py:123]: Based on how we are setting `noita_regions` below, I don't think we need any empty sets at all, as long as the region is used anywhere at all in this dict.

**Comment by black-sliver** [comment]
The amount of (empty) sets is concerning :sweat_smile: 

Looking at the code, I think the connections could (and should) be lists. Lists are cheaper, unless you use the `in` operator, and are ordered.

Following that code, i think `noita_regions` should be a list as well. Maybe
`noita_regions: List[str] = sorted(set(noita_connections.keys()).union(*noita_connections.values()))`
to remove duplicates and have stable order.

**Comment by ScipioWright** [comment]
Ah, I always thought sets were faster generally, and that lists were for if you needed something ordered. I've changed it over to lists.

**Review by heinermann — APPROVED** [review]

**Comment by black-sliver** [comment]
> I always thought sets were faster generally

Iteration over sets is "fine". It's not gonna be faster than tuple or list. But also iteration over a list or tuple is deterministic while it isn't for sets, so using sets may get you in trouble depending on what you do.
`in` for sets is (up to A LOT) faster if you have more than ~2 members.
set always uses more memory than list or tuple, so it's more resource intensive (if it doesn't give you an advantage), and thus it's easier to get cache misses, making it slower in some circumstances


Just double-checking. The final version of this was tested to some extend?

**Comment by ScipioWright** [comment]
That's correct. I've tested it by doing several gens (some with one yaml, some with 10, some with other games), and loading into a max settings game and making sure stuff was where I expected it to be.

**Review by black-sliver — APPROVED** [review]

**Pull request merged**